### ISSUE TITLE
DynamoDB add custom resources

### ISSUE THREAD

**Issue by jagi** [serverless/serverless#1699]
Hi, I'm new to Serverless and I want to use DynamoDB with severless@1.0-alpha. As there is no event source for DynamoDB then I have to add custom resources provider. It has change how we define it from 0.x version. It used to use JSON and now it's using YAML. However I see in docs that it's still possible to reference JSON file and here is what I have.

**cloudformation-resources.json**
```json
{
  "AWSTemplateFormatVersion": "2010-09-09",
  "Description": "",
  "Resources": {
    "IamPolicyLambdaInvocationDynamoDB": {
      "Type": "AWS::IAM::Policy",
      "Properties": {
        "PolicyName": "${stage}-${project}-dynamo-stream-lambda",
        "PolicyDocument": {
          "Version": "2012-10-17",
          "Statement": [{
            "Effect": "Allow",
            "Action": [
              "lambda:InvokeFunction"
            ],
            "Resource": "*"
          }, {
            "Effect": "Allow",
            "Action": [
              "dynamodb:DescribeStream",
              "dynamodb:GetRecords",
              "dynamodb:GetShardIterator",
              "dynamodb:ListStreams",
              "dynamodb:PutItem",
              "dynamodb:GetItem"
            ],
            "Resource": "*"
          }]
        },
        "Roles": [{
          "Ref": "IamRoleLambda"
        }]
      }
    }
  }
}
```
**serverless.yaml**
```yaml
service: aws-nodejs
provider: aws
runtime: nodejs4.3

functions:
  hello:
    handler: handler.hello
    events:
      - http:
          path: greet
          method: get

resources:
  Resources:
    $ref: ./cloudformation-resources.json
```

When I try to run this function I get the following error `Under Resources block, the resources have to be in JSON format. For debugging logs, run again after setting SLS_DEBUG env var.` What's wrong with that? Right now I have to add permissions by hand in the AWS console.

**Comment by eahefnawy** [comment]
Thanks for reporting this. To get more insight about this, could you please set the `SLS_DEBUG` env var by running `export SLS_DEBUG=*` and run again and send us the output/stack trace?

You also may want to try the unreleased `v1.0` branch which is up to date and may have fixed a related issue.

**Comment by kevinold** [comment]
@eahefnawy I'm seeing the same issue and here is the stack trace I receive:

```
ValidationError: Under Resources block, the resources have to be in JSON format
    at Request.extractError (/Users/kold/github_clones/serverless/node_modules/aws-sdk/lib/protocol/query.js:40:29)
    at Request.callListeners (/Users/kold/github_clones/serverless/node_modules/aws-sdk/lib/sequential_executor.js:105:20)
    at Request.emit (/Users/kold/github_clones/serverless/node_modules/aws-sdk/lib/sequential_executor.js:77:10)
    at Request.emit (/Users/kold/github_clones/serverless/node_modules/aws-sdk/lib/request.js:615:14)
    at Request.transition (/Users/kold/github_clones/serverless/node_modules/aws-sdk/lib/request.js:22:10)
    at AcceptorStateMachine.runTo (/Users/kold/github_clones/serverless/node_modules/aws-sdk/lib/state_machine.js:14:12)
    at /Users/kold/github_clones/serverless/node_modules/aws-sdk/lib/state_machine.js:26:10
    at Request.<anonymous> (/Users/kold/github_clones/serverless/node_modules/aws-sdk/lib/request.js:38:9)
    at Request.<anonymous> (/Users/kold/github_clones/%
```

**Comment by ac360** [comment]
@kevinold is back!!!

**Comment by kevinold** [comment]
@ac360 Indeed! :D 

I bypassed the JSON route and ended up getting my cloudformation-resources.json which was valid json, converted to yaml by using https://www.npmjs.com/package/yamljs and the json2yaml utility it provides.

My resulting resources block looks like so:

```
# you can add CloudFormation resource templates here
 resources:
     Resources:
       IamRoleLambda: {Type: 'AWS::IAM::Role', Properties: {AssumeRolePolicyDocument: {Version: '2012-10-17', Statement: [{Effect: Allow, Pr  incipal: {Service: [lambda.amazonaws.com]}, Action: ['sts:AssumeRole']}]}, Path: /}}
       IamPolicyLambda: {Type: 'AWS::IAM::Policy', Properties: {PolicyName: '${stage}-${service}-lambda', PolicyDocument: {Version: '2012-10  -17', Statement: [{Effect: Allow, Action: ['logs:CreateLogGroup', 'logs:CreateLogStream', 'logs:PutLogEvents'], Resource: 'arn:aws:logs:${r  egion}:*:*'}, {Effect: Allow, Action: ['*'], Resource: 'arn:aws:dynamodb:${region}:*:table/${service}-posts-${stage}'}, {Effect: Allow, Act  ion: ['*'], Resource: 'arn:aws:dynamodb:${region}:*:table/${service}-authors-${stage}'}, {Effect: Allow, Action: ['*'], Resource: 'arn:aws:  dynamodb:${region}:*:table/${service}-comments-${stage}'}]}, Roles: [{Ref: IamRoleLambda}]}}
       PostsDynamo: {Type: 'AWS::DynamoDB::Table', DeletionPolicy: Retain, Properties: {AttributeDefinitions: [{AttributeName: id, Attribute  Type: S}], KeySchema: [{AttributeName: id, KeyType: HASH}], ProvisionedThroughput: {ReadCapacityUnits: 1, WriteCapacityUnits: 1}, TableName  : '${service}-posts-${stage}'}}
       AuthorsDynamo: {Type: 'AWS::DynamoDB::Table', DeletionPolicy: Retain, Properties: {AttributeDefinitions: [{AttributeName: id, Attribu  teType: S}], KeySchema: [{AttributeName: id, KeyType: HASH}], ProvisionedThroughput: {ReadCapacityUnits: 1, WriteCapacityUnits: 1}, TableNa  me: '${service}-authors-${stage}'}}
       CommentsDynamo: {Type: 'AWS::DynamoDB::Table', DeletionPolicy: Retain, Properties: {AttributeDefinitions: [{AttributeName: id, Attrib  uteType: S}], KeySchema: [{AttributeName: id, KeyType: HASH}], ProvisionedThroughput: {ReadCapacityUnits: 1, WriteCapacityUnits: 1}, TableN  ame: '${service}-comments-${stage}'}}
     Outputs:
       IamRoleArnLambda: {Description: 'ARN of the lambda IAM role', Value: {'Fn::GetAtt': [IamRoleLambda, Arn]}}
```

Note that this is the https://github.com/serverless/serverless-graphql-blog/blob/master/s-resources-cf.json converted to yaml.

As you can see from the yaml, I removed these sections from the top of the JSON for it to work in `serverless.yaml`

```
"AWSTemplateFormatVersion": "2010-09-09",
  "Description": "The AWS CloudFormation template for this Serverless application's resources outside of Lambdas and Api Gateway",
  "Resources": {
```

I had no issues deploying my DynamoDB table's with this yaml resource config

**Comment by kevinold** [comment]
Also, I'm waiting on #1711 to be merged, and added `service` and `region` to my `serverless.env.yaml` to complete the namespacing on deploy.

**Comment by kevinold** [comment]
I found a more readable solution:

`json2yaml cloudformation-resources.json --depth 8` produces:

```
AWSTemplateFormatVersion: '2010-09-09'
Description: 'The AWS CloudFormation template for this Serverless application''s resources outside of Lambdas and Api Gateway'
Resources:
  IamRoleLambda:
    Type: 'AWS::IAM::Role'
    Properties:
      AssumeRolePolicyDocument:
        Version: '2012-10-17'
        Statement:
          -
            Effect: Allow
            Principal:
              Service: [lambda.amazonaws.com]
            Action:
              - 'sts:AssumeRole'
      Path: /
  IamPolicyLambda:
    Type: 'AWS::IAM::Policy'
    Properties:
      PolicyName: '${stage}-${project}-lambda'
      PolicyDocument:
        Version: '2012-10-17'
        Statement:
          -
            Effect: Allow
            Action:
              - 'logs:CreateLogGroup'
              - 'logs:CreateLogStream'
              - 'logs:PutLogEvents'
            Resource: 'arn:aws:logs:${region}:*:*'
          -
            Effect: Allow
            Action:
              - '*'
            Resource: 'arn:aws:dynamodb:${region}:*:table/${project}-posts-${stage}'
          -
            Effect: Allow
            Action:
              - '*'
            Resource: 'arn:aws:dynamodb:${region}:*:table/${project}-authors-${stage}'
          -
            Effect: Allow
            Action:
              - '*'
            Resource: 'arn:aws:dynamodb:${region}:*:table/${project}-comments-${stage}'
      Roles:
        -
          Ref: IamRoleLambda
  PostsDynamo:
    Type: 'AWS::DynamoDB::Table'
    DeletionPolicy: Retain
    Properties:
      AttributeDefinitions:
        -
          AttributeName: id
          AttributeType: S
      KeySchema:
        -
          AttributeName: id
          KeyType: HASH
      ProvisionedThroughput:
        ReadCapacityUnits: 1
        WriteCapacityUnits: 1
      TableName: '${project}-posts-${stage}'
  AuthorsDynamo:
    Type: 'AWS::DynamoDB::Table'
    DeletionPolicy: Retain
    Properties:
      AttributeDefinitions:
        -
          AttributeName: id
          AttributeType: S
      KeySchema:
        -
          AttributeName: id
          KeyType: HASH
      ProvisionedThroughput:
        ReadCapacityUnits: 1
        WriteCapacityUnits: 1
      TableName: '${project}-authors-${stage}'
  CommentsDynamo:
    Type: 'AWS::DynamoDB::Table'
    DeletionPolicy: Retain
    Properties:
      AttributeDefinitions:
        -
          AttributeName: id
          AttributeType: S
      KeySchema:
        -
          AttributeName: id
          KeyType: HASH
      ProvisionedThroughput:
        ReadCapacityUnits: 1
        WriteCapacityUnits: 1
      TableName: '${project}-comments-${stage}'
Outputs:
  IamRoleArnLambda:
    Description: 'ARN of the lambda IAM role'
    Value:
      'Fn::GetAtt':
        - IamRoleLambda
        - Arn
```

Remove the top 2 lines and you can drop right into your `serverless.yaml`

**Comment by jagi** [comment]
With the `SLS_DEBUG` set

```
Validation Error ---------------------------------------
 
     Under Resources block, the resources have to be in JSON
     format
 
     For debugging logs, run again after setting SLS_DEBUG env var.
 
  Stack Trace --------------------------------------------
 
ValidationError: Under Resources block, the resources have to be in JSON format
    at Request.extractError (/Users/jagi/.nvm/versions/node/v4.3.2/lib/node_modules/serverless/node_modules/aws-sdk/lib/protocol/query.js:40:29)
    at Request.callListeners (/Users/jagi/.nvm/versions/node/v4.3.2/lib/node_modules/serverless/node_modules/aws-sdk/lib/sequential_executor.js:105:20)
    at Request.emit (/Users/jagi/.nvm/versions/node/v4.3.2/lib/node_modules/serverless/node_modules/aws-sdk/lib/sequential_executor.js:77:10)
    at Request.emit (/Users/jagi/.nvm/versions/node/v4.3.2/lib/node_modules/serverless/node_modules/aws-sdk/lib/request.js:615:14)
    at Request.transition (/Users/jagi/.nvm/versions/node/v4.3.2/lib/node_modules/serverless/node_modules/aws-sdk/lib/request.js:22:10)
    at AcceptorStateMachine.runTo (/Users/jagi/.nvm/versions/node/v4.3.2/lib/node_modules/serverless/node_modules/aws-sdk/lib/state_machine.js:14:12)
    at /Users/jagi/.nvm/versions/node/v4.3.2/lib/node_modules/serverless/node_modules/aws-sdk/lib/state_machine.js:26:10
    at Request.<anonymous> (/Users/jagi/.nvm/versions/node/v4.3.2/lib/node_modules/serverless/node_modules/aws-sdk/lib/request.js:38:9)
    at Request.<anonymous> (/Users/jagi/.nvm/versions/node/v4.3.2/lib/node_modules/serverless/node_modules/aws-sdk/lib/request.js:617:12)
    at Request.callListeners (/Users/jagi/.nvm/versions/node/v4.3.2/lib/node_modules/serverless/node_modules/aws-sdk/lib/sequential_executor.js:115:18)
    at Request.emit (/Users/jagi/.nvm/versions/node/v4.3.2/lib/node_modules/serverless/node_modules/aws-sdk/lib/sequential_executor.js:77:10)
    at Request.emit (/Users/jagi/.nvm/versions/node/v4.3.2/lib/node_modules/serverless/node_modules/aws-sdk/lib/request.js:615:14)
    at Request.transition (/Users/jagi/.nvm/versions/node/v4.3.2/lib/node_modules/serverless/node_modules/aws-sdk/lib/request.js:22:10)
    at AcceptorStateMachine.runTo (/Users/jagi/.nvm/versions/node/v4.3.2/lib/node_modules/serverless/node_modules/aws-sdk/lib/state_machine.js:14:12)
    at /Users/jagi/.nvm/versions/node/v4.3.2/lib/node_modules/serverless/node_modules/aws-sdk/lib/state_machine.js:26:10
    at Request.<anonymous> (/Users/jagi/.nvm/versions/node/v4.3.2/lib/node_modules/serverless/node_modules/aws-sdk/lib/request.js:38:9)
 
  Get Support --------------------------------------------
     Docs:          v1.docs.serverless.com
     Bugs:          github.com/serverless/serverless/issues
 
     Please report this error. We think it might be a bug.
```

**Comment by eahefnawy** [comment]
ahaaaaa of course!! Thanks for the pointer @kevinold ... I haven't noticed that @jagi `cloudformation-resources.json` file contains the `AWSTemplateFormatVersion` & `Description` properties.

@jagi our docs are probably misleading on that front, we're really sorry about that 😞 . `cloudformation-resources.json` file **is NOT meant to be a full CloudFormation template file**, but rather a JSON or YAML file that holds a bunch of **Resource Templates**.

So to fix the issue, you should just take all of your resource templates inside that `Resources` property, and put them all at the highest level of the JSON file, and remove the `AWSTemplateFormatVersion` & `Description` properties.

So basically that means we only support adding `Resources` for CloudFormation, but you can't add other main properties like `Outputs` for example. In the future we may extend that support with something like:

```yml
resources:
  Resources:
    $ref: ./file1.json
  Outputs:
    $ref: ./file2.json
```

I hope this helps 😊

**Comment by jagi** [comment]
@eahefnawy it works thanks :)
